core/spec/models/spree/product/scopes_spec.rb

```diff
@@ -183,4 +183,28 @@
       end
     end
   end
+
+  describe '.with_variant_sku_cont' do
+    let!(:product) { create(:product, name: 'ring cap') }
+    let!(:variant_1) { create(:variant, product: product, sku: 'H100') }
+    let!(:variant_2) { create(:variant, product: product, sku: 'H290') }
+
+    context 'query products' do
+      it 'scope returns empty array' do
+        product.discard
+        expect(Spree::Product.with_variant_sku_cont('H1').to_a).to eq []
+      end
+
+      it 'scope returns product' do
+        product = Spree::Product.with_variant_sku_cont('H1').first
+        expect(product.name).to eq 'ring cap'
+      end
+
+      it 'returns product when with_discarded is passed on product' do
+        product.discard
+        product = Spree::Product.with_discarded.with_variant_sku_cont('H1').first
+        expect(product.name).to eq 'ring cap'
+      end
+    end
+  end
```

Comment:
Thanks for adding these specs. What if we also discard the variant instead? Like:

```ruby
result_with_variant_not_discarded = Spree::Product.with_variant_sku_cont('H1').to_a
variant_1.discard
result_with_variant_discarded = Spree::Product.with_variant_sku_cont('H1').to_a
expect(result_with_variant_not_discarded).to eq result_with_variant_discarded
```

I think this expectation should fail and it doesn't with the current code otherwise we have no way of deciding if we want to return discarded variants or not. What do you think?

Reply:
yes, it should fail, since the corresponding product isn't discarded. But since the scope is only used within the context of returning products, deleted or not,  with a specific `SKU`; I think that should be the expected behaviour. I might be wrong on this. When we discard a product it also discards its corresponding variants. If we need to search for that product with a `SKU` which has been discarded, we will not get anything back. So I would propose to maybe change this method's name to reflect that it is for searching variants discarded or not; Or we can create a new scope for that purpose.

Reply:
> When we discard a product it also discards its corresponding variants. 

But it's not true the other way around. Discarding a variant without discarding the whole product is totally legit.

My concern here is that the search form will return products with discarded variants (not the whole product) searching for the SKU of a discarded variant, even when `Show deleted` is not selected. Can you please verify?

On the technical implementation, I was also thinking at another scope that explicitly tells that we will return also discarded variants, like `.with_all_variant_sku_cont` and maybe we can rename `with_variant_sku_cont` to `with_kept_variant_sku_cont` to reflect discard gem's terminology.

Reply:
I have looked into it by discarding a variant of an active(non-discarded) product. 
`Index Action`
It will return that product with that `SKU` even if the variant is discarded. This is the default behaviour whether or not you check the `Show deleted` checkbox and without invoking the `with_variant_sku_cont` scope.
`Search Functionality`
with invoking the `with_variant_sku_cont` scope. It will also return that product with the discarded variant.

So either way, if we discard a variant, but a product is linked to that `SKU`, that variant will be returned. Regardless if we are using the `search functionality` or just the `index`

Reply:
> I would suggest to move the SKU to the product model and make it unique since SKUs should be unique to a specific product. That way we will solve both problems.

I don't think we can do that, each variant has a different SKU.

Reply:
Then I think it leaves us with few options: 
`first`: to leave the implementation as it is
`second`: to return products that only have undiscarded variances
`Third`: to find a middle ground between the two. (we may need to brainstorm on this one)
 what do you think?

Reply:
What about creating a new scope called `with_all_variant_sku_cont` that also check for discarded variants and connect it to the search when `Show deleted` is checked? When `Show deleted` is not selected we'll keep using the `with_variant_sku_cont` scope. Do you think this solution can work or am I missing something?

Reply:
yes, it can work. But discarded variants will still be available visible as long as the product is still linked to it.

Reply:
The highlighted product is active but its variance is discarded. This is the behavior I was hinting about. 

<img width="1221" alt="Screen Shot 2021-02-05 at 9 35 41 AM" src="https://user-images.githubusercontent.com/23207591/106998536-b76b9080-6795-11eb-9de0-0e9a8b772416.png">

Reply:
Products can have multiple variants. I think that we can consider product not discarded with only one variant (called master variant) discarded as an edge case. In real life you discard a variant of a product when you have at least another one non discarded. Can you try to see if it make sense discarding a variant of a product that has multiple variants?

Reply:
okay, let me check it out.

Reply:
The search_form and index action only return master variants. When you delete a different variance of the same product and search it with the old `with_variance_sku_cont` scope it will return that product. The same can be said after adding in my changes. It seems like when searching priority is given to master variance. 
Would it be possible if we can schedule a video call because this looks quite big of a problem

Reply:
I have been reviewing your proposed options for adding a scope to explicitly search discarded variances and I think that would be the best approach.